hedera-node/src/main/java/com/hedera/services/state/migration/MigrationRecordsManager.java

```diff
@@ -224,13 +270,91 @@ private void publishContractFreeAutoRenewalRecords() {
+                                        "Published synthetic bytecode sidecar for contract 0.0.{}",
+                                        contractId.getContractNum());
+                                // create state changes if contract has storage
+                                var contractStorageKey = account.getFirstContractStorageKey();
+                                if (contractStorageKey == null) {
+                                    log.debug(
+                                            "Contract 0.0.{} has no iterable storage - no state"
+                                                    + " changes will be published.",
+                                            contractId.getContractNum());
+                                } else {
+                                    final var contractStateChangeBuilder =
+                                            ContractStateChange.newBuilder()
+                                                    .setContractId(contractId);
+                                    IterableContractValue iterableValue;
+                                    while (contractStorageKey != null) {
+                                        iterableValue = contractStorageMap.get(contractStorageKey);
+                                        contractStateChangeBuilder.addStorageChanges(
+                                                StorageChange.newBuilder()
+                                                        .setSlot(
+                                                                ByteStringUtils.wrapUnsafely(
+                                                                        slotAsBytes(
+                                                                                contractStorageKey)))
+                                                        .setValueRead(
+                                                                ByteStringUtils.wrapUnsafely(
+                                                                        iterableValue.getValue()))
+                                                        .build());
+                                        contractStorageKey =
+                                                iterableValue.getNextKeyScopedTo(
+                                                        contractStorageKey.getContractId());
+                                    }
+                                    allContractsStateChangesBuilder.addContractStateChanges(
+                                            contractStateChangeBuilder);
+                                    log.debug(
+                                            "Published synthetic state changes for contract 0.0.{}",
+                                            contractId.getContractNum());
+                                }
+                            }
+                        });
+        if (allContractsStateChangesBuilder.getContractStateChangesCount() > 0) {
+            transactionContext.addSidecarRecord(
```

Comment:
Am I correct in assuming we're creating a single sidecar record with all the state changes from all the contracts existing on the network? What if the record is larger than 256MB or the size limit for the record files? Will the splitting algorithm be able to split it into multiple files?
The alternative would be to create a record for each contract and add all the records to the transaction context.

Reply:
Yeah, it will definitely be a problem if the contract storage in the network > `sidecarMaxSize` and the single `TransactionSidecarRecord` for the state changes becomes too large (maybe can happen if we perform the migration on preview/test net, not sure of the stats there?). We haven't really handled the case in the sidecar file limit logic when a **single** `TransactionSidecarRecord` exceeds the max file size and I'm not sure whether we should 🤔 

IMO creating a separate `TransactionSidecarRecord` for each contract or at least grouping them into smaller batches is a better approach. If we go in this direction maybe we should confirm that the mirror node won't have a problem with multiple `TransactionSidecarRecords` with state changes having the same `consensus_timestamp.` (I guess this applies to bytecodes as well).

Reply:
For migrations, we should not put everything in one sidecar file and should still respect the 256MB limit and split it across multiple files. The mirror node can support multiple `TransactionSidecarRecord` for the same contract since the unique key for state change is "timestamp, contract id, slot".

For bytecode migration, the timestamp doesn't matter since we just need the contract ID to associate the bytecode.

Reply:
Okay, modified now to create a separate state changes sidecar for each contract, instead of putting all the state changes in a single sidecar.